src/vm/method.cpp

```diff
@@ -4768,15 +4809,40 @@ Precode* MethodDesc::GetOrCreatePrecode()
         return GetPrecode();
     }
 
-    PTR_PCODE pSlot = GetAddrOfSlot();
-    PCODE tempEntry = GetTemporaryEntryPoint();
-
     PrecodeType requiredType = GetPrecodeType();
     PrecodeType availableType = PRECODE_INVALID;
 
-    if (!GetMethodDescChunk()->HasCompactEntryPoints())
+    PTR_PCODE pSlot = GetAddrOfSlot();
+    PCODE pExpected;
+#ifdef FEATURE_PREJIT
+    if (IsZapped())
+    {
+        // If the method is zapped, then the only way it may not have a stable entry point yet
+        // is that it has a precode. In this case it's not marked as having precode through HasPrecode
+        // instead the zapped precode acts as a temporary entry point.
+        pExpected = *pSlot;
```

Comment:
I know we probably do this in other places ... but this pattern is prone to be broken by C/C++ compiler optimizations.

https://github.com/dotnet/coreclr/pull/16089 is a recent example 

The fix is to use VolatileLoadWithoutBarrier to load the value. Like: `pExpected = VolatileLoadWithoutBarrier(pSlot);`